File: examples/python/01_unique_items.py
import json

import oc_sidememory

from _transformers import accepts, load_runtime, masked_next_token


model, tokenizer, vocabulary = load_runtime()
schema = json.dumps({
    "type": "array",
    "items": {"type": "string", "enum": ["a", "b"]},
    "minItems": 2,
    "maxItems": 2,
    "uniqueItems": True,
})
guide = oc_sidememory.SidememoryGuide(
    oc_sidememory.compile_schema(schema, vocabulary, tokenizer.vocab_size)
)
assert accepts(guide, tokenizer, '["a","b"]')
assert not accepts(guide, tokenizer, '["a","a"]')
selected = masked_next_token(model, tokenizer, guide)
print(f"uniqueItems: real-model token {selected!r}; duplicate rejected")

File: examples/rust/01_unique_items.rs
mod common;

use oc_sidememory::sidememory::ProbeDecision;

fn main() {
    let schema = r#"{"type":"array","items":{"type":"string","enum":["a","b"]},"minItems":2,"maxItems":2,"uniqueItems":true}"#;
    let compiled = common::compile(schema, &[br#"["a","b"]"#, br#"["a","a"]"#], None);
    let mut guide = common::guide(compiled);
    assert_eq!(guide.probe(0).expect("probe succeeds"), ProbeDecision::Allow);
    assert!(matches!(guide.probe(1).expect("probe succeeds"), ProbeDecision::Deny(_)));
    println!("uniqueItems: accepted distinct values and rejected a duplicate");
}
